test/e2e/e2e_test.go

```diff
@@ -70,9 +89,77 @@ func TestMain(m *testing.M) {
 		cancel()
 	}()
 
+	if globalOpts.ArtifactDir != "" {
+		go setupMetricsEndpoint(testContext, log)
+		go e2eObserverControllers(testContext, log, globalOpts.ArtifactDir)
+		defer dumpTestMetrics(log, globalOpts.ArtifactDir)
+	}
+
 	// Everything's okay to run tests
 	log.Info("executing e2e tests", "options", globalOpts)
-	os.Exit(m.Run())
+	return m.Run()
+}
+
+func e2eObserverControllers(ctx context.Context, log logr.Logger, artifactDir string) {
+	mgr, err := ctrl.NewManager(e2eutil.GetConfigOrDie(), manager.Options{MetricsBindAddress: "0"})
+	if err != nil {
+		log.Error(err, "failed to construct manager for observers")
+		return
+	}
+	if err := podtimingcontroller.SetupWithManager(mgr, log, artifactDir); err != nil {
+		log.Error(err, "failed to set up podtimingcontroller")
+		return
+	}
+
+	if err := mgr.Start(ctx); err != nil {
+		log.Error(err, "Mgr ended")
+	}
+}
+
+const metricsServerAddr = "127.0.0.1:8080"
+
+func setupMetricsEndpoint(ctx context.Context, log logr.Logger) {
+	log.Info("Setting up metrics endpoint")
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	server := &http.Server{Addr: metricsServerAddr, Handler: mux}
+
+	go func() {
+		<-ctx.Done()
+		if err := server.Shutdown(ctx); err != nil {
+			log.Error(err, "Shutting down metrics server failed")
+		}
+	}()
+	if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		log.Error(err, "metrics server ended unexpectedly")
+	}
+}
+
+func dumpTestMetrics(log logr.Logger, artifactDir string) {
+	log.Info("Fetching test metrics")
+	response, err := http.Get(metricsServerAddr + "/metrics")
```

Comment:
URL is missing a scheme, probably accounts for this error in the CI logs:

```
time="2021-09-28T16:37:49Z" level=error msg="error fetching test metrics" error="parse \"127.0.0.1:8080/metrics\": first path segment in URL cannot contain colon"
```

Reply:
Fixed